src/plot.cpp

```diff
@@ -980,16 +1019,552 @@ RGBColor random_color(void)
+   * since the stack size varies within each column.
+   */
+#ifdef _OPENMP
+  const int n_threads = omp_get_max_threads();
+#else
+  const int n_threads = 1;
+#endif
+  std::vector<std::vector<std::vector<TrackSegment>>> this_line_segments(
+    n_threads);
+  for (int t = 0; t < n_threads; ++t) {
+    this_line_segments[t].resize(pixels_[0]);
+  }
+
+  // The last thread writes to this, and the first thread reads from it.
+  std::vector<std::vector<TrackSegment>> old_segments(pixels_[0]);
+
+#pragma omp parallel
+  {
+
+#ifdef _OPENMP
+    const int n_threads = omp_get_max_threads();
+    const int tid = omp_get_thread_num();
+#else
+    int n_threads = 1;
+    int tid = 0;
+#endif
+
+    SourceSite s; // Where particle starts from (camera)
+    s.E = 1;
+    s.wgt = 1;
+    s.delayed_group = 0;
+    s.particle = ParticleType::photon; // just has to be something reasonable
+    s.parent_id = 1;
+    s.progeny_id = 2;
+    s.r = camera_position_;
+
+    Particle p;
+    s.u.x = 1.0;
+    s.u.y = 0.0;
+    s.u.z = 0.0;
+    p.from_source(&s);
+
+    int vert = tid;
+    for (int iter = 0; iter <= pixels_[1] / n_threads; iter++) {
+
+      // Save bottom line of current work chunk to compare against later
+      // I used to have this inside the below if block, but it causes a
+      // spurious line to be drawn at the bottom of the image. Not sure
+      // why, but moving it here fixes things.
+      if (tid == n_threads - 1)
+        old_segments = this_line_segments[n_threads - 1];
+
+      if (vert < pixels_[1]) {
+
+        for (int horiz = 0; horiz < pixels_[0]; ++horiz) {
+
+          // Generate the starting position/direction of the ray
+          if (orthographic_width_ == 0.0) { // perspective projection
+            double this_phi = -horiz_fov_radians / 2.0 + dphi * horiz;
+            double this_mu = -vert_fov_radians / 2.0 + dmu * vert + M_PI / 2.0;
+            Direction camera_local_vec;
+            camera_local_vec.x = std::cos(this_phi) * std::sin(this_mu);
+            camera_local_vec.y = std::sin(this_phi) * std::sin(this_mu);
+            camera_local_vec.z = std::cos(this_mu);
+            s.u = camera_local_vec.rotate(camera_to_model);
+          } else { // orthographic projection
+            s.u = looking_direction;
+
+            double x_pix_coord = (static_cast<double>(horiz) - p0 / 2.0) / p0;
+            double y_pix_coord = (static_cast<double>(vert) - p1 / 2.0) / p0;
+            s.r = camera_position_ +
+                  cam_yaxis * x_pix_coord * orthographic_width_ +
+                  cam_zaxis * y_pix_coord * orthographic_width_;
+          }
+
+          p.from_source(&s); // put particle at camera
+          bool hitsomething = false;
+          bool intersection_found = true;
+          int loop_counter = 0;
+          const int max_intersections = 1000000;
+
+          this_line_segments[tid][horiz].clear();
+
+          int first_surface =
```

Comment:
It seems like factoring out the first intersection would allow you to get rid of some of this state that's only used in the first case in the loop.

Reply:
Also, is it possible to exit and re-enter the model once the particle is inside?

Reply:
Yeah, I forgot about this assumption I made! Are OpenMC models assumed to be convex? If we're simulating radiation transport, it seems like a bad model to ever not be convex as it's clearly physically incorrect: you'd be artificially killing particles that may be streaming into some other place you'd like to tally.

Maybe there is some exception to this where reflective BCs would be employed, but I can't imagine it right now. Maybe it would make sense to note in the documentation that the model must be convex rather than allowing this edge case? I think the logic might be complicated to handle reentrant geometries.

Reply:
That's a good point about the model itself. We don't disallow non-convex model boundaries, but it certainly could cause some modeling problems you've identified here. Moreso I'm wondering if there are CSG cells inside the model will we get the correct answer? I suppose so because we're using the same logic for particles undergoing transport and assuming we're already inside a cell for the ray tracing logic.